File: src/main/java/io/github/mdsimmo/bomberman/messaging/Contexted.java
package io.github.mdsimmo.bomberman.messaging;

import org.bukkit.command.CommandSender;
import org.bukkit.inventory.ItemStack;

import javax.annotation.CheckReturnValue;
import javax.annotation.Nonnull;
import java.util.Collection;

@CheckReturnValue
@Nonnull
public interface Contexted {

    @Nonnull Contexted with(String key, Formattable arg);

    @Nonnull Message format();

    @Nonnull
    default Contexted with(@Nonnull String key, @Nonnull String value) {
        return with(key, new Formattable.StringWrapper(value));
    }

    @Nonnull
    default Contexted with(String key, int value) {
        return with(key, new Formattable.StringWrapper(String.valueOf(value)));
    }

    default Contexted with(String key, Collection<? extends Formattable> value) {
        return with(key, new Formattable.CollectionWrapper<>(value));
    }

    default Contexted with(String key, ItemStack stack) {
        return with(key, new Formattable.ItemWrapper(stack));
    }

    default Contexted with(String key, CommandSender sender) {
        return with(key, new Formattable.SenderWrapper(sender));
    }

    default void sendTo(CommandSender sender) {
        format().sendTo(sender);
    }

}

File: src/main/java/io/github/mdsimmo/bomberman/messaging/Contexted.kt
package io.github.mdsimmo.bomberman.messaging

import org.bukkit.command.CommandSender
import org.bukkit.inventory.ItemStack
import javax.annotation.CheckReturnValue


/**
 * A Contexted is a string that can have Formattables attached to it and then expanded into a message
 */
@CheckReturnValue
interface Contexted {

    fun with(key: String, thing: Formattable): Contexted

    fun format(): Message

    fun with(key: String, value: String): Contexted {
        return with(key, StringWrapper(value))
    }

    fun with(key: String, value: Int): Contexted {
        return with(key, StringWrapper(value.toString()))
    }

    fun with(key: String, value: Collection<Formattable>): Contexted {
        return with(key, CollectionWrapper(value))
    }

    fun with(key: String, stack: ItemStack): Contexted {
        return with(key, ItemWrapper(stack))
    }

    fun with(key: String, sender: CommandSender): Contexted {
        return with(key, SenderWrapper(sender))
    }

    fun sendTo(sender: CommandSender) {
        format().sendTo(sender)
    }
}